src/rgw/rgw_rest_s3.cc

```diff
@@ -4208,16 +4244,23 @@ void rgw::auth::s3::LDAPEngine::shutdown() {
   }
 }
 
-/* LocalEndgine */
+/* LocalEngine */
 rgw::auth::Engine::result_t
 rgw::auth::s3::LocalEngine::authenticate(
   const boost::string_view& _access_key_id,
   const boost::string_view& signature,
+  const boost::string_view& session_token,
   const string_to_sign_t& string_to_sign,
   const signature_factory_t& signature_factory,
   const completer_factory_t& completer_factory,
   const req_state* const s) const
 {
+  //If LocalAuth is enabled and the request has a session token, then LocalEngine shouldn't authenticate it.
+  if (s->info.args.exists("X-Amz-Security-Token") ||
+      s->info.env->exists("HTTP_X_AMZ_SECURITY_TOKEN")) {
+    return result_t::deny();
```

Comment:
it feels awkward that LDAP/LocalEngine need to know about sts tokens. can we find a better way to get this behavior?

for example, could we run the STSEngine before the others? if no sts token is given, it can return a `result_t::deny` and the strategy will go on and try the other engines. but if an sts token _is_ given and fails to authenticate, it can return `result_t::reject` to skip the other engines

would that make sense? would it break anything else?

Reply:
If lets say the user calls GetSessionToken, the authentication of which will be using Keystone and then the next call will be to other s3 apis using the credentials returned in the previous step, the authentication of this step will be using STS. In this case, the config file will have both keystone_auth and sts_auth enabled, so if STS is inserted before all engines, in case of GetSessionToken, result_t::reject will result in Keystone engine being skipped. result::deny may work, but I will have to think it through once.

Reply:
i see, thanks. would it make sense to create a separate `AuthStrategy` for these `RGWRESTMgr_STS` ops that only does keystone? then if `AWSAuthStrategy` is only used for `RGWRESTMgr_S3`, it should be okay for that strategy to put the sts engine first and use `result_t::reject` as above?

Reply:
Sounds like this will work, I will try it out.

Reply:
@cbodley : I have changed the code to add STS engine before other engines, and the STS engine returns a deny in case Session Token is not part of the incoming request, and returns a reject for all other auth failures. The last two commits are related to this change, can you please re-review?